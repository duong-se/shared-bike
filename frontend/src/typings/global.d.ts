export type Config = {
  appEnv: 'development' | 'production'
  baseUrl: string
  googleMapApiKey: string
}

declare global {
  interface Window {
    sharedBike: {
      config: Config
    }
    google: typeof google
  }
}
