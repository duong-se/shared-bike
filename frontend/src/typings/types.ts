export type LoginVariables = {
  username: string
  password: string
}

export type LoginResponse = {
  accessToken: string
}

export type User = {
  id: number
  name: string
  username: string
}

export type RentBikeVariables = {
  bikeId: number
}

export type ReturnBikeVariables = {
  bikeId: number
}

export enum BikeStatus {
  RENTED = "rented",
  AVAILABLE = "available"
}

export type Bike = {
  id: number
  name: string
  lat: string
  long: string
  status: BikeStatus
  userId?: number
  nameOfRenter?: string
  usernameOfRenter?: string
}

export type RegisterVariables = {
  username: string
  password: string
  name: string
}

export type RegisterResponse = {
  accessToken: string
}
