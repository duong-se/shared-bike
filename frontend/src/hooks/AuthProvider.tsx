import React, { useEffect } from "react"
import jwtDecode from 'jwt-decode'
import { tokenKey } from "../constants/constants";
import { User } from "../typings/types";


type AuthContextType = {
  setUser: React.Dispatch<React.SetStateAction<User | undefined>>
  user?: User
}

const AuthContext = React.createContext<AuthContextType>(null!);

export const useAuth = () => {
  return React.useContext(AuthContext);
}

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [user, setUser] = React.useState<User|undefined>()
  useEffect(() => {
    const token = localStorage.getItem(tokenKey)
    if (token) {
      const decoded = jwtDecode<User>(token);
      setUser(decoded);
    }
  }, [])
  return <AuthContext.Provider value={{ user, setUser }}>{children}</AuthContext.Provider>;
}
