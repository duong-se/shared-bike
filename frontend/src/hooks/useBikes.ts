import { useMutation, UseMutationOptions, useQuery, UseQueryOptions } from "react-query"
import { fetchBikes, rentBike, returnBike } from "../apis/bikes"
import { Bike, RentBikeVariables, ReturnBikeVariables } from "../typings/types"

export const useBikes = (options?: Omit<UseQueryOptions<Bike[], unknown, Bike[], string[]>, "queryKey" | "queryFn">) => {
  return useQuery(['bikes'], fetchBikes, options)
}

export const useRentBike = (options?: Omit<UseMutationOptions<Bike, unknown, RentBikeVariables, unknown>, "mutationFn">) => {
  return useMutation(rentBike, options)
}

export const useReturnBike = (options?: Omit<UseMutationOptions<Bike, unknown, ReturnBikeVariables, unknown>, "mutationFn">) => {
  return useMutation(returnBike, options)
}
