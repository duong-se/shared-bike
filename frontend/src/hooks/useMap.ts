import { useCallback, useState } from 'react'

export type UseMapReturn = {
  mapRef: (node: HTMLDivElement | null) => void
  map?: google.maps.Map
}

export const useMap = (position: google.maps.LatLng): UseMapReturn => {
  const [map, setMap] = useState<google.maps.Map>()
  const mapRef = useCallback((node: HTMLDivElement | null) => {
    if (node !== null) {
      const map = new window.google.maps.Map(node, {
        center: position,
        zoom: 8,
        fullscreenControl: false,
      })
      setMap(map)
    }
  }, [])
  return {
    map,
    mapRef,
  }
}
