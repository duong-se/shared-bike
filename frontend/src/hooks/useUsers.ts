import { useMutation, UseMutationOptions } from 'react-query'
import { login, register } from '../apis/users'
import { LoginResponse, LoginVariables } from '../typings/types'
import { RegisterResponse, RegisterVariables } from '../typings/types'

export const useLogin = (options?: Omit<UseMutationOptions<LoginResponse, unknown, LoginVariables, unknown>, 'mutationFn'>) => {
  return useMutation(login, options)
}

export const useRegister = (options?: Omit<UseMutationOptions<RegisterResponse, unknown, RegisterVariables, unknown>, 'mutationFn'>) => {
  return useMutation(register, options)
}
