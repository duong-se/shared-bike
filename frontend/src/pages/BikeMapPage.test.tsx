import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from 'react-query'
import * as reactQuery from 'react-query'
import { initialize, mockInstances, Map, Marker, InfoWindow } from '@googlemaps/jest-mocks'
import {
  availableColor,
  BikeMapPage,
  customIcon,
  rentedColor,
  contentMap,
  renderCommonContent,
  renderUserHasBikeCase,
  renderUserHasNoBikeCase,
  onRentOrReturnSuccess,
  handleMarkerCallback,
  handlePopUpButtonCallback,
} from './BikeMapPage'
import * as useBikes from '../hooks/useBikes'
import { Bike, BikeStatus } from '../typings/types'
import * as useAuth from '../hooks/AuthProvider'
import * as useMap from '../hooks/useMap'

describe('BikeMapPage', () => {
  beforeEach(() => {
    initialize()
  })
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  const Provider: React.FC<React.PropsWithChildren> = ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
  it('should render loading', async () => {
    jest.spyOn(reactQuery, 'useQueryClient').mockReturnValue({
      setQueryData: jest.fn(),
    } as unknown as reactQuery.QueryClient)
    jest.spyOn(useAuth, 'useAuth').mockReturnValue({
      user: {
        id: 1,
        name: 'mockName',
        username: 'mockUsername',
      },
      setUser: jest.fn()
    })
    jest.spyOn(useBikes, 'useBikes').mockReturnValue({
      isLoading: true,
      isSuccess: false,
      data: [{
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      }]
    } as unknown as reactQuery.UseQueryResult<Bike[], unknown>)
    jest.spyOn(useMap, 'useMap').mockReturnValue({
      map: mockInstances.get(Map)[0] as unknown as google.maps.Map,
      mapRef: jest.fn()
    })

    render(<Provider><BikeMapPage /></Provider>)
    await waitFor(() => {
      const spinner = screen.getByRole('status')
      expect(spinner).toBeInTheDocument()
    })
  })

  it('should render map when user have bike', async () => {
    jest.spyOn(reactQuery, 'useQueryClient').mockReturnValue({
      setQueryData: jest.fn(),
    } as unknown as reactQuery.QueryClient)
    jest.spyOn(useAuth, 'useAuth').mockReturnValue({
      user: {
        id: 1,
        name: 'mockName',
        username: 'mockUsername',
      },
      setUser: jest.fn()
    })
    jest.spyOn(useBikes, 'useBikes').mockReturnValue({
      isLoading: false,
      isSuccess: true,
      data: [{
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      }]
    } as unknown as reactQuery.UseQueryResult<Bike[], unknown>)
    jest.spyOn(useMap, 'useMap').mockReturnValue({
      map: mockInstances.get(Map)[0] as unknown as google.maps.Map,
      mapRef: jest.fn()
    })

    render(<Provider><BikeMapPage /></Provider>)
    await waitFor(() => {
      const map = screen.getByTestId('map-container')
      expect(map).toBeInTheDocument()
    })
  })

  it('should render map when user have no bike', async () => {
    jest.spyOn(reactQuery, 'useQueryClient').mockReturnValue({
      setQueryData: jest.fn(),
    } as unknown as reactQuery.QueryClient)
    jest.spyOn(useAuth, 'useAuth').mockReturnValue({
      user: {
        id: 1,
        name: 'mockName',
        username: 'mockUsername',
      },
      setUser: jest.fn()
    })
    jest.spyOn(useBikes, 'useBikes').mockReturnValue({
      isLoading: false,
      isSuccess: true,
      data: [{
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.AVAILABLE,
        nameOfRenter: '',
        userId: 0,
      }]
    } as unknown as reactQuery.UseQueryResult<Bike[], unknown>)
    jest.spyOn(useMap, 'useMap').mockReturnValue({
      map: mockInstances.get(Map)[0] as unknown as google.maps.Map,
      mapRef: jest.fn()
    })

    render(<Provider><BikeMapPage /></Provider>)
    await waitFor(() => {
      const map = screen.getByTestId('map-container')
      expect(map).toBeInTheDocument()
    })
  })

  it('should render error', async () => {
    jest.spyOn(reactQuery, 'useQueryClient').mockReturnValue({
      setQueryData: jest.fn(),
    } as unknown as reactQuery.QueryClient)
    jest.spyOn(useAuth, 'useAuth').mockReturnValue({
      user: {
        id: 1,
        name: 'mockName',
        username: 'mockUsername',
      },
      setUser: jest.fn()
    })
    jest.spyOn(useBikes, 'useBikes').mockReturnValue({
      isLoading: false,
      isSuccess: true,
      isError: true,
      error: 'mockError',
      data: null
    } as unknown as reactQuery.UseQueryResult<Bike[], unknown>)
    jest.spyOn(useMap, 'useMap').mockReturnValue({
      map: mockInstances.get(Map)[0] as unknown as google.maps.Map,
      mapRef: jest.fn()
    })

    render(<Provider><BikeMapPage /></Provider>)
    const error = (await screen.findByText('mockError')).textContent
    expect(error).toEqual('mockError')
  })
})

describe('customIcon', () => {
  it('should render custom icon with available color', () => {
    const icon = customIcon(availableColor)
    expect(icon).toMatchSnapshot()
  })
  it('should render custom icon with rented color', () => {
    const icon = customIcon(rentedColor)
    expect(icon).toMatchSnapshot()
  })
})

describe('contentMap', () => {
  it('should render content map for available bike', () => {
    const map = contentMap[BikeStatus.AVAILABLE]
    const view = map.renderButton({
      id: 1,
      name: 'mockName',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    } as Bike)
    expect(view).toContain('bike-action-1')
    expect(view).toContain('RENT BIKE')
  })

  it('should render content map for rented bike', () => {
    const map = contentMap[BikeStatus.RENTED]
    const view = map.renderButton({
      id: 1,
      name: 'mockName',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    } as Bike)
    expect(view).toContain('bike-action-1')
    expect(view).toContain('RENT BIKE')
  })

  it('should render content map for retur bike', () => {
    const map = contentMap.returnBike
    const view = map.renderButton({
      id: 1,
      name: 'mockName',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    } as Bike)
    expect(view).toContain('bike-action-1')
    expect(view).toContain('RETURN BIKE')
  })
})

describe('renderCommonContent', () => {
  it('should renderCommonContent', () => {
    const view = renderCommonContent({
      id: 1,
      name: 'mockName',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    } as Bike)
    expect(view).toContain('mockName')
  })
})

describe('renderUserHasBikeCase', () => {
  beforeEach(() => {
    initialize()
  })
  it('should render bikes', () => {
    const mockReturnBikeMutate = jest.fn()
    const map = new google.maps.Map(document.createElement('div'))
    const view = renderUserHasBikeCase(
      [
        {
          id: 1,
          name: 'mockName',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter',
          userId: 1,
        },
        {
          id: 2,
          name: 'mockName1',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter1',
          userId: 1,
        },
      ] as Array<Bike>,
      {
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      },
      mockReturnBikeMutate,
      map,
    )
    const infoWindowMocks = mockInstances.get(InfoWindow)
    const markerMocks = mockInstances.get(Marker)
    expect(markerMocks).toHaveLength(2)
    expect(markerMocks[0].setIcon).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[0].setContent).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[0].addListener).toHaveBeenCalledTimes(1)
    expect(markerMocks[0].addListener).toHaveBeenCalledTimes(1)
    expect(markerMocks[1].setIcon).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[1].setContent).toHaveBeenCalledTimes(1)
    expect(markerMocks[1].addListener).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[1].addListener).toHaveBeenCalledTimes(0)
    expect(view).toHaveLength(2)
  })

  it('should return market empty', () => {
    const mockReturnBikeMutate = jest.fn()
    const view = renderUserHasBikeCase(
      [
        {
          id: 1,
          name: 'mockName',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter',
          userId: 1,
        },
        {
          id: 2,
          name: 'mockName1',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter1',
          userId: 1,
        },
      ] as Array<Bike>,
      {
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      },
      mockReturnBikeMutate,
      undefined,
    )
    expect(view).toHaveLength(0)
  })
})


describe('renderUserHasNoBikeCase', () => {
  beforeEach(() => {
    initialize()
  })
  it('should render bikes', () => {
    const mockReturnBikeMutate = jest.fn()
    const map = new google.maps.Map(document.createElement('div'))
    const view = renderUserHasNoBikeCase(
      [
        {
          id: 1,
          name: 'mockName',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter',
          userId: 1,
        },
        {
          id: 2,
          name: 'mockName1',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.AVAILABLE,
          nameOfRenter: '',
          userId: 0,
        },
      ] as Array<Bike>,
      mockReturnBikeMutate,
      map,
    )
    const infoWindowMocks = mockInstances.get(InfoWindow)
    const markerMocks = mockInstances.get(Marker)
    expect(markerMocks).toHaveLength(2)
    expect(markerMocks[0].setIcon).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[0].addListener).toHaveBeenCalledTimes(0)
    expect(markerMocks[0].addListener).toHaveBeenCalledTimes(1)
    expect(markerMocks[1].setIcon).toHaveBeenCalledTimes(1)
    expect(markerMocks[1].addListener).toHaveBeenCalledTimes(1)
    expect(infoWindowMocks[1].addListener).toHaveBeenCalledTimes(1)
    expect(view).toHaveLength(2)
  })

  it('should return empty', () => {
    const mockReturnBikeMutate = jest.fn()
    const view = renderUserHasNoBikeCase(
      [
        {
          id: 1,
          name: 'mockName',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter',
          userId: 1,
        },
        {
          id: 2,
          name: 'mockName1',
          lat: '50.123456',
          long: '8.123456',
          status: BikeStatus.RENTED,
          nameOfRenter: 'mockRenter1',
          userId: 1,
        },
      ] as Array<Bike>,
      mockReturnBikeMutate,
      undefined,
    )
    expect(view).toHaveLength(0)
  })
})

describe('onRentOrReturnSuccess', () => {
  it('should run success', () => {
    const queryClient = new QueryClient()
    const spy = jest.spyOn(queryClient, 'setQueryData').mockReturnValue([
      {
        id: 1,
        name: 'mockName1',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      },
      {
        id: 2,
        name: 'mockName1',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter1',
        userId: 1,
      }
    ])
    jest.spyOn(queryClient, 'getQueryData').mockReturnValue([
      {
        id: 1,
        name: 'mockName',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      },
      {
        id: 2,
        name: 'mockName1',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter1',
        userId: 1,
      },
    ])
    const fn = onRentOrReturnSuccess(queryClient)
    fn({
      id: 1,
      name: 'mockName1',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    })
    expect(spy).toBeCalled()
  })

  it('should run for no currentBike', () => {
    const queryClient = new QueryClient()
    const spy = jest.spyOn(queryClient, 'setQueryData').mockReturnValue([
      {
        id: 1,
        name: 'mockName1',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter',
        userId: 1,
      },
      {
        id: 2,
        name: 'mockName1',
        lat: '50.123456',
        long: '8.123456',
        status: BikeStatus.RENTED,
        nameOfRenter: 'mockRenter1',
        userId: 1,
      }
    ])
    jest.spyOn(queryClient, 'getQueryData').mockReturnValue(undefined)
    const fn = onRentOrReturnSuccess(queryClient)
    fn({
      id: 1,
      name: 'mockName1',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    })
    expect(spy).not.toBeCalled()
  })
})

describe('handleMarkerCallback', () => {
  beforeEach(() => {
    initialize()
  })
  it('should run and call call open fn', () => {
    const map = new google.maps.Map(document.createElement('div'))
    const infoWindow = new google.maps.InfoWindow()
    const marker = new google.maps.Marker()
    const fn = handleMarkerCallback(infoWindow, marker, map)
    fn()
    const infoWindowMocks = mockInstances.get(InfoWindow)
    expect(infoWindowMocks[0].open).toHaveBeenCalledTimes(1)
  })
})

describe('handlePopUpButtonCallback', () => {
  const MockComponent = () => <button data-testid="test-button-1" id="bike-action-1">mock</button>
  beforeEach(() => {
    initialize()
  })
  it('should run and call open fn', () => {
    const infoWindow = new google.maps.InfoWindow()
    const mockFn = jest.fn()
    const mockBike = {
      id: 1,
      name: 'mockName1',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    }
    render(<MockComponent />)
    const fn = handlePopUpButtonCallback(mockFn, mockBike, infoWindow)
    fn()
    const markerButton = screen.getByTestId('test-button-1')
    fireEvent.click(markerButton)
    const infoWindowMocks = mockInstances.get(InfoWindow)
    expect(mockFn).toBeCalled()
    expect(infoWindowMocks[0].close).toHaveBeenCalledTimes(1)
  })

  it('should run and cannot find button', () => {
    const infoWindow = new google.maps.InfoWindow()
    const mockFn = jest.fn()
    const mockBike = {
      id: 1,
      name: 'mockName1',
      lat: '50.123456',
      long: '8.123456',
      status: BikeStatus.RENTED,
      nameOfRenter: 'mockRenter',
      userId: 1,
    }
    render(<div />)
    const fn = handlePopUpButtonCallback(mockFn, mockBike, infoWindow)
    fn()
  })
})
