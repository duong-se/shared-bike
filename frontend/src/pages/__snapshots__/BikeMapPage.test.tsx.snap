// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customIcon should render custom icon with available color 1`] = `
Object {
  "fillColor": "#2ecc71",
  "fillOpacity": 1,
  "path": "M 0,0 C -2,-20 -10,-22 -10,-30 A 10,10 0 1,1 10,-30 C 10,-22 2,-20 0,0 z M -2,-30 a 2,2 0 1,1 4,0 2,2 0 1,1 -4,0",
  "scale": 1,
  "strokeColor": "#000",
  "strokeWeight": 2,
}
`;

exports[`customIcon should render custom icon with rented color 1`] = `
Object {
  "fillColor": "#34495e",
  "fillOpacity": 1,
  "path": "M 0,0 C -2,-20 -10,-22 -10,-30 A 10,10 0 1,1 10,-30 C 10,-22 2,-20 0,0 z M -2,-30 a 2,2 0 1,1 4,0 2,2 0 1,1 -4,0",
  "scale": 1,
  "strokeColor": "#000",
  "strokeWeight": 2,
}
`;
